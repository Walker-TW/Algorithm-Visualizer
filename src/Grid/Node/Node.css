.Node {
  width: 20px;
  height: 20px;
  border: 1px solid black;
  background-color: lightcyan;
}

.visited {
  background-color: hotpink;
}

.start {
  background-color: rgb(206, 47, 255);
}

.finish {
  background-color: yellow;
}

.path {
  background-color: chartreuse;
}
