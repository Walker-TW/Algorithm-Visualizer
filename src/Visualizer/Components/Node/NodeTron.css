.Node {
  width: 20px;
  height: 20px;
  border: 2px solid rgb(189, 189, 5);
  /* background-color: black; */
  /* opacity: 0.5; */
  transition: transform 0.2s;
  border-radius: 2px;
}

@media (max-width: 1030px) {
  .Node {
    width: 30px;
    height: 30px;
  }
}

@media (max-width: 500px) {
  .Node {
    width: 40px;
    height: 40px;
  }
}

.passage {
  background-color: olive;
}
.visited {
  animation-name: visitedAnimation;
  animation-duration: 0.3s;
  animation-timing-function: ease-out;
  animation-delay: 0;
  animation-direction: alternate;
  animation-iteration-count: 1;
  animation-fill-mode: forwards;
  animation-play-state: running;
}
@keyframes visitedAnimation {
  0% {
    transform: scale(0.3);
    background-color: rgb(255, 4, 100);
  }

  50% {
    transform: scale(1.3);
    background-color: rgb(218, 157, 27);
  }

  100% {
    transform: scale(1);
    background-color: rgb(255, 225, 0);
    border: 1px solid black;
  }
}

.start {
  background-color: rgb(206, 47, 255);
  animation-name: startAnimation;
  animation-duration: 0.3s;
  animation-timing-function: ease-out;
  animation-delay: 0;
  animation-direction: alternate;
  animation-iteration-count: 1;
  animation-fill-mode: forwards;
  animation-play-state: running;
}
@keyframes startAnimation {
  0% {
    transform: scale(0.3);
    background-color: rgb(23, 154, 45);
  }

  50% {
    transform: scale(1.2);
    background-color: rgb(194, 142, 28);
  }

  100% {
    transform: scale(1);
    background-color: rgb(206, 47, 255);
  }
}

.finish {
  background-color: yellow;
  animation-name: finishAnimation;
  animation-duration: 0.3s;
  animation-timing-function: ease-out;
  animation-delay: 0;
  animation-direction: alternate;
  animation-iteration-count: 1;
  animation-fill-mode: forwards;
  animation-play-state: running;
}
@keyframes finishAnimation {
  0% {
    transform: scale(0.3);
    background-color: rgb(23, 154, 45);
  }

  50% {
    transform: scale(1.2);
    background-color: rgb(194, 142, 28);
  }

  100% {
    transform: scale(1);
    background-color: yellow;
  }
}

.path {
  animation-name: pathAnimation;
  animation-duration: 0.3s;
  animation-timing-function: ease-out;
  animation-delay: 0;
  animation-direction: alternate;
  animation-iteration-count: 1;
  animation-fill-mode: forwards;
  animation-play-state: running;
}
@keyframes pathAnimation {
  0% {
    transform: scale(0.3);
    background-color: rgb(186, 24, 178);
  }

  50% {
    transform: scale(1.2);
    background-color: rgb(247, 191, 70);
  }

  100% {
    transform: scale(1);
    background-color: rgb(0, 174, 255);
  }
}

.fence {
  /* background-color: orange; */
  animation-name: fenceAnimation;
  animation-duration: 0.1s;
  animation-timing-function: ease-out;
  animation-delay: 0;
  animation-direction: alternate;
  animation-iteration-count: 1;
  animation-fill-mode: forwards;
  animation-play-state: running;
}

@keyframes fenceAnimation {
  /* 0% {
    transform: scale(0.3);
    background-color: rgb(23, 154, 45);
  } */

  0% {
    transform: scale(1.3);
    background-color: rgb(194, 142, 28);
  }

  100% {
    transform: scale(1);
    background-color: rgb(255, 0, 0);
  }
}
