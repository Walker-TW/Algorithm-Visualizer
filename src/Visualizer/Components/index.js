export * from './Grid';
export * from './Node';
export * from './Header';
export * from './Stats';
