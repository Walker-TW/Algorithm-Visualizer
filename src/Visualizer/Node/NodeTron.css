.Node {
  width: 20px;
  height: 20px;
  border: 2px solid rgb(189, 189, 5);
  /* background-color: black; */
  /* opacity: 0.5; */
  transition: transform 0.2s;
  border-radius: 2px;
}

@media (max-width: 900px) {
  .Node {
    width: 30px;
    height: 30px;
  }
}

@media (max-width: 500px) {
  .Node {
    width: 40px;
    height: 40px;
  }
}

.visited {
  animation-name: visitedAnimation;
  animation-duration: 0.3s;
  animation-timing-function: ease-out;
  animation-delay: 0;
  animation-direction: alternate;
  animation-iteration-count: 1;
  animation-fill-mode: forwards;
  animation-play-state: running;
}
@keyframes visitedAnimation {
  0% {
    transform: scale(0.3);
    background-color: rgb(19, 169, 129);
  }

  /* 15% {
    transform: scale(0.7);
    background-color: rgb(22, 186, 63);
  } */
  /* 25% {
    transform: scale(1);
    background-color: rgb(51, 225, 25);
  } */

  50% {
    transform: scale(1.3);
    background-color: rgb(218, 157, 27);
  }

  /* 75% {
    transform: scale(1.1);
    background-color: rgb(189, 114, 0);
  } */

  /* 80% {
    transform: scale(1.2);
    background-color: rgb(145, 128, 2);
  } */

  100% {
    transform: scale(1);
    /* background: radial-gradient(circle, yellow, #9b9ba3, #090979); */
    background-color: rgb(255, 225, 0);
    border: 1px solid black;
  }
}

.start {
  background-color: rgb(206, 47, 255);
  animation-name: startAnimation;
  animation-duration: 0.3s;
  animation-timing-function: ease-out;
  animation-delay: 0;
  animation-direction: alternate;
  animation-iteration-count: 1;
  animation-fill-mode: forwards;
  animation-play-state: running;
}
@keyframes startAnimation {
  0% {
    transform: scale(0.3);
    background-color: rgb(23, 154, 45);
  }

  50% {
    transform: scale(1.2);
    background-color: rgb(194, 142, 28);
  }

  100% {
    transform: scale(1);
    background-color: rgb(206, 47, 255);
  }
}

.finish {
  background-color: yellow;
  animation-name: finishAnimation;
  animation-duration: 0.3s;
  animation-timing-function: ease-out;
  animation-delay: 0;
  animation-direction: alternate;
  animation-iteration-count: 1;
  animation-fill-mode: forwards;
  animation-play-state: running;
}
@keyframes finishAnimation {
  0% {
    transform: scale(0.3);
    background-color: rgb(23, 154, 45);
  }

  50% {
    transform: scale(1.2);
    background-color: rgb(194, 142, 28);
  }

  100% {
    transform: scale(1);
    background-color: yellow;
  }
}

.path {
  background-color: chartreuse;
  animation-name: pathAnimation;
  animation-duration: 0.3s;
  animation-timing-function: ease-out;
  animation-delay: 0;
  animation-direction: alternate;
  animation-iteration-count: 1;
  animation-fill-mode: forwards;
  animation-play-state: running;
}
@keyframes pathAnimation {
  0% {
    transform: scale(0.3);
    background-color: rgb(186, 24, 178);
  }

  50% {
    transform: scale(1.2);
    background-color: rgb(247, 191, 70);
  }

  100% {
    transform: scale(1);
    background-color: chartreuse;
  }
}

.fence {
  animation-name: fenceAnimation;
  animation-duration: 0.3s;
  animation-timing-function: ease-out;
  animation-delay: 0;
  animation-direction: alternate;
  animation-iteration-count: 1;
  animation-fill-mode: forwards;
  animation-play-state: running;
}

@keyframes fenceAnimation {
  0% {
    transform: scale(0.3);
    background-color: rgb(23, 154, 45);
  }

  50% {
    transform: scale(1.2);
    background-color: rgb(194, 142, 28);
  }

  100% {
    transform: scale(1);
    background-color: rgb(255, 0, 0);
  }
}
