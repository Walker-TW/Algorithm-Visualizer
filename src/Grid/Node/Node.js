import React, { Component } from "react";
import "./Node.css";
import PropTypes from "prop-types";

export default class Node extends Component {
  state = {
    start: false,
    finish: false,
    fence: false,
  };

  clickHandler = () => {
    const {
      gridId,
      gridHasStart,
      gridHasFinish,
      gridHasFenceToggle,
      nodeFlag,
    } = this.props;

    if (gridHasStart && gridHasFinish && !gridHasFenceToggle) {
      document.getElementById("button").className = "btn btn-primary big";
    } else {
      if (!gridHasStart) {
        nodeFlag(gridId, "start");
        this.setState({ start: true });
      } else {
        this.setState({ finish: true });
        nodeFlag(gridId, "finish");
      }
    }
  };

  fenceSelector = (e) => {
    const { gridId, nodeFlag } = this.props;
    const { fence } = this.state;
    if (fence === false) {
      nodeFlag(gridId, "fence");
      this.setState({ fence: true });
    } else {
      nodeFlag(gridId, "fence");
      this.setState({ fence: false });
    }
  };

  mouseUpHandler = () => {
    const { mouseFlag, mouseToggle, gridHasFenceToggle } = this.props;
    if (gridHasFenceToggle === true && mouseToggle === true) {
      mouseFlag();
    }
  };
  mouseEnterHandler = () => {
    const { mouseToggle, gridHasFenceToggle } = this.props;
    if (gridHasFenceToggle === true && mouseToggle === true) {
      this.fenceSelector();
    }
  };
  mouseDownHandler = () => {
    const { mouseToggle, mouseFlag, gridHasFenceToggle } = this.props;
    if (gridHasFenceToggle === true && mouseToggle === false) {
      mouseFlag();
      this.fenceSelector();
    } else {
      this.clickHandler();
    }
  };

  render() {
    return (
      <div
        onMouseDown={this.mouseDownHandler}
        onMouseEnter={this.mouseEnterHandler}
        onMouseUp={this.mouseUpHandler}
        className={`Node ${
          this.state.start
            ? "start"
            : this.state.finish
            ? "finish"
            : this.state.fence
            ? "fence"
            : ""
        }`}
        id={this.props.id}
      />
    );
  }
}

Node.propTypes = {
  gridId: PropTypes.object.isRequired,
  nodeFlag: PropTypes.func.isRequired,
  gridHasStart: PropTypes.bool.isRequired,
  gridHasFinish: PropTypes.bool.isRequired,
  gridHasFenceToggle: PropTypes.bool.isRequired,

  // reset: PropTypes.func.isRequired,
};
