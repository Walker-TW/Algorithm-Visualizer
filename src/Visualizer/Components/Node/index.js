import React, { Component } from 'react';
import PropTypes from 'prop-types';

// import "./Node.css";
import './NodeTron.css';
// import './AnimatedNode.css';
class Node extends Component {
  state = {
    start: false,
    finish: false,
    fence: false,
  };

  // start and end points are not adjustable
  // at the moment so they have a seperate handler

  startFinishHandler = (type) => {
    const { gridId, nodeFlag } = this.props;
    nodeFlag(gridId, [type]);
    this.setState({ [type]: true });
  };

  fenceHandler = () => {
    const { gridId, nodeFlag } = this.props;
    const { fence, start, finish } = this.state;

    if (!start && !finish) nodeFlag(gridId, 'fence');
    this.setState({ fence: !fence });
  };

  mouseUpHandler = () => {
    const { mouseFlag, mouseToggle, fenceToggle } = this.props;

    if (fenceToggle && mouseToggle) mouseFlag();
  };

  mouseEnterHandler = () => {
    const { mouseToggle, fenceToggle } = this.props;

    if (fenceToggle && mouseToggle) this.fenceHandler();
  };

  mouseDownHandler = () => {
    const { mouseToggle, mouseFlag, fenceToggle } = this.props;
    if (fenceToggle && !mouseToggle) {
      mouseFlag();
      this.fenceHandler();
    } else {
      this.clickHandler();
    }
  };

  clickHandler = () => {
    const { gridHasStart, gridHasFinish, resetStartFinish } = this.props;
    const { start, finish } = this.state;

    if (gridHasStart && gridHasFinish) {
      // visual aid
      document.getElementById('run-btn').innerText = 'Click Me!';

      let reset = start ? 'start' : finish ? 'finish' : null;

      if (start || finish) resetStartFinish(reset);
    } else if (!gridHasStart) {
      if (finish) alert("Start and end can't be the same");
      else this.startFinishHandler('start');
    } else if (gridHasStart && !gridHasFinish) {
      if (start) {
        alert("Start and end can't be the same");
      } else {
        this.startFinishHandler('finish');
      }
    }
  };

  render() {
    const { start, finish, fence } = this.state;
    const { size } = this.props;
    return (
      <div
        onMouseDown={this.mouseDownHandler}
        onMouseEnter={this.mouseEnterHandler}
        onMouseUp={this.mouseUpHandler}
        style={
          size === 0 ? null : { width: parseInt(size), height: parseInt(size) }
        }
        className={`Node ${
          start ? 'start' : finish ? 'finish' : fence ? 'fence' : ''
        }`}
        id={this.props.id}
      />
    );
  }
}

export { Node };

Node.propTypes = {
  fenceToggle: PropTypes.bool.isRequired,
  gridId: PropTypes.object.isRequired,
  gridHasStart: PropTypes.bool.isRequired,
  gridHasFinish: PropTypes.bool.isRequired,
  mouseFlag: PropTypes.func.isRequired,
  mouseToggle: PropTypes.bool.isRequired,
  nodeFlag: PropTypes.func.isRequired,
  resetStartFinish: PropTypes.func.isRequired,
  size: PropTypes.number.isRequired,
};
