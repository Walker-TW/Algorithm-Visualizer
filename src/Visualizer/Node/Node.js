import React, { Component } from 'react';
import PropTypes from 'prop-types';

// import "./Node.css";
import './NodeTron.css';
// import './AnimatedNode.css';
export default class Node extends Component {
  state = {
    start: false,
    finish: false,
    fence: false,
  };

  // start and end points are not adjustable
  // at the moment so they have a seperate handler

  startFinishHandler = (type) => {
    const { gridId, nodeFlag } = this.props;
    nodeFlag(gridId, [type]);
    this.setState({ [type]: true });
  };

  fenceHandler = () => {
    const { gridId, nodeFlag } = this.props;
    const { fence, start, finish } = this.state;

    if (!start && !finish) nodeFlag(gridId, 'fence');
    this.setState({ fence: !fence });
  };

  mouseUpHandler = () => {
    const { mouseFlag, mouseToggle, fenceToggle } = this.props;

    if (fenceToggle && mouseToggle) mouseFlag();
  };

  mouseEnterHandler = () => {
    const { mouseToggle, fenceToggle } = this.props;

    if (fenceToggle && mouseToggle) this.fenceHandler();
  };

  mouseDownHandler = () => {
    const { mouseToggle, mouseFlag, fenceToggle } = this.props;
    if (fenceToggle && !mouseToggle) {
      mouseFlag();
      this.fenceHandler();
    } else {
      this.clickHandler();
    }
  };

  clickHandler = () => {
    const { gridHasStart, gridHasFinish, resetStartFinish } = this.props;
    const { start, finish } = this.state;

    if (gridHasStart && gridHasFinish) {
      // visual aid
      document.getElementById('run-btn').innerText = 'Click Me!';
      if (start) resetStartFinish('start');
      if (finish) resetStartFinish('finish');
    }
    if (!gridHasStart) {
      if (finish) {
        alert("Start and end can't be the same");
      } else {
        this.startFinishHandler('start');
      }
    }
    if (gridHasStart && !gridHasFinish) {
      if (start) {
        alert("Start and end can't be the same");
      } else {
        this.startFinishHandler('finish');
      }
    }
  };

  render() {
    const { start, finish, fence } = this.state;
    return (
      <div
        onMouseDown={this.mouseDownHandler}
        onMouseEnter={this.mouseEnterHandler}
        onMouseUp={this.mouseUpHandler}
        className={`Node ${
          start ? 'start' : finish ? 'finish' : fence ? 'fence' : ''
        }`}
        id={this.props.id}
      />
    );
  }
}

Node.propTypes = {
  gridId: PropTypes.object.isRequired,
  gridHasStart: PropTypes.bool.isRequired,
  gridHasFinish: PropTypes.bool.isRequired,
  nodeFlag: PropTypes.func.isRequired,
  mouseFlag: PropTypes.func.isRequired,
  mouseToggle: PropTypes.bool.isRequired,
  fenceToggle: PropTypes.bool.isRequired,
  resetStartFinish: PropTypes.func.isRequired,
};
