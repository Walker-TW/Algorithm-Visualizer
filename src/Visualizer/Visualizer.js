import React, { Component, Fragment } from 'react';

import Header from './Header/Header';

import Alert from 'react-bootstrap/Alert';

import Nodes from './Nodes/Nodes';

import { dijkstra, findShortestPath } from '../Algorithms/PathFinders/dijkstra';
import {
  aStarManhatten,
  findShortestPathAStarM,
} from '../Algorithms/PathFinders/a*manhatten';
import {
  aStarEuclidean,
  findShortestPathAStarE,
} from '../Algorithms/PathFinders/a*euclidean';
import {
  breadthFirstSearch,
  findShortestPathBFS,
} from '../Algorithms/PathFinders/bfs.js';
import {
  depthFirstSearch,
  findShortestPathDFS,
} from '../Algorithms/PathFinders/dfs.js';
import './Visualizer.css';

export default class Visualizer extends Component {
  state = {
    algorithm: '',
    grid: [],
    fenceToggle: false,
    mouseToggle: false,
    start: {
      present: false,
      gridId: {
        rowIndex: null,
        colIndex: null,
      },
    },
    finish: {
      present: false,
      gridId: {
        rowIndex: null,
        colIndex: null,
      },
      fence: {
        present: [false],
        gridId: {
          rowIndex: null,
          colIndex: null,
        },
      },
    },
  };

  // setup methods
  componentDidMount() {
    this.gridSetup();
    window.addEventListener('resize', this.gridSetup);
  }
  componentWillUnmount() {
    window.removeEventListener('resize', this.gridSetup);
  }

  createNode = (gridId) => {
    return {
      gridId,
      start: false,
      finish: false,
      visited: false,
      pastNode: null,
      fence: false,
      heuristic: Infinity,
      manhatten: Infinity,
      distance: Infinity,
    };
  };

  // getDimensions = () => {
  //   const [screenWidth, screenHeight] = [window.innerWidth, window.innerHeight];

  //   let width, height;
  //   if (screenWidth > 1450) {
  //     width = screenWidth / 20 - 12;
  //     height = screenHeight / 20 - 7;
  //   } else if (screenWidth > 900) {
  //     width = screenWidth / 30;
  //     height = screenHeight / 30;
  //   } else {
  //     width = screenWidth / 40 - 3;
  //     height = screenHeight / 40 - 5;
  //   }
  //   return [width, height];
  // };

  gridSetup = () => {
    const [screenWidth, screenHeight] = [window.innerWidth, window.innerHeight];

    let width, height;
    if (screenWidth > 1450) {
      width = screenWidth / 20 - 12;
      height = screenHeight / 20 - 7;
    } else if (screenWidth > 900) {
      width = screenWidth / 30;
      height = screenHeight / 30;
    } else {
      width = screenWidth / 40 - 3;
      height = screenHeight / 40 - 5;
    }

    let grid = [];
    for (let rowIndex = 0; rowIndex < width; rowIndex++) {
      let current_row = [];
      for (let colIndex = 0; colIndex < height; colIndex++) {
        const gridId = { colIndex, rowIndex };

        current_row.push(this.createNode(gridId));
      }
      grid.push(current_row);
    }
    this.setState({ grid });
  };

  // state changers & prop methods
  fenceToggler = () => {
    const { fenceToggle } = this.state;

    this.setState({ fenceToggle: !fenceToggle });
  };

  mouseFlag = () => {
    const { mouseToggle } = this.state;

    this.setState({ mouseToggle: !mouseToggle });
  };

  nodeFlag = (gridId, type) => {
    const { grid } = this.state;
    const { rowIndex, colIndex } = gridId;
    const node = grid[rowIndex][colIndex];
    if (type === 'fence') {
      node[type] = !node[type];
    } else {
      node[type] = true;
      this.setState({
        [type]: {
          present: true,
          gridId: gridId,
        },
      });
    }
  };

  setAlgorithm = (selection) => {
    if (this.state.algorithm !== selection)
      this.setState({ algorithm: selection });
  };

  reset = () => {
    let grid = [...this.state.grid];
    grid.forEach((row) => {
      row.forEach((node) => {
        this.resetNodeHandler(node);
      });
    });

    this.setState({ grid: grid });
  };

  resetNodeHandler = (node) => {
    const resetNodeStyle = (node) => {
      document.getElementById(
        `node-${node.gridId.colIndex}-${node.gridId.rowIndex}`
      ).className = `Node`;
    };

    const resetNode = (node) => {
      node.heuristic = Infinity;
      node.manhatten = Infinity;
      node.distance = Infinity;
      node.visited = false;
      node.pastNode = null;
    };

    if (node.start || node.fence || node.finish) {
      resetNode(node);
    } else {
      // resetNode(node);
      // this has been added because the nodes were not being created so instead they are resetted
      resetNodeStyle(node);
      node = this.createNode(node.gridId);
    }
  };

  run = () => {
    const { algorithm } = this.state;

    if (algorithm === 'dijkstra') {
      this.runDijkstra();
    } else if (algorithm === 'A* Euclidean') {
      this.runAstarEuclidean();
    } else if (algorithm === 'A* Manhatten') {
      this.runAstarManhatten();
    } else if (algorithm === 'Depth First Search') {
      this.runDepthFirstSearch();
    } else if (algorithm === 'Breadth First Search') {
      this.runBreadthFirstSearch();
    }
  };

  runBreadthFirstSearch = () => {
    const { grid, start, finish } = this.state;
    const startNode = grid[start.gridId.rowIndex][start.gridId.colIndex];
    const finishNode = grid[finish.gridId.rowIndex][finish.gridId.colIndex];
    const resultOfBreadthFirstSearch = breadthFirstSearch(
      grid,
      startNode,
      finishNode
    );
    const z = findShortestPathBFS(
      resultOfBreadthFirstSearch[resultOfBreadthFirstSearch.length - 1]
    );
    this.animateAlgorithm(resultOfBreadthFirstSearch, z);
  };

  runDepthFirstSearch = () => {
    const { grid, start, finish } = this.state;
    const startNode = grid[start.gridId.rowIndex][start.gridId.colIndex];
    const finishNode = grid[finish.gridId.rowIndex][finish.gridId.colIndex];
    const resultOfDepthFirstSearch = depthFirstSearch(
      grid,
      startNode,
      finishNode
    );
    const z = findShortestPathDFS(
      resultOfDepthFirstSearch[resultOfDepthFirstSearch.length - 1]
    );
    this.animateAlgorithm(resultOfDepthFirstSearch, z);
  };

  runAstarEuclidean = () => {
    const { grid, start, finish } = this.state;
    const startNode = grid[start.gridId.rowIndex][start.gridId.colIndex];
    const finishNode = grid[finish.gridId.rowIndex][finish.gridId.colIndex];
    const resultOfAStarE = aStarEuclidean(grid, startNode, finishNode);
    const y = findShortestPathAStarE(resultOfAStarE[resultOfAStarE.length - 1]);
    this.animateAlgorithm(resultOfAStarE, y);
  };

  runAstarManhatten = () => {
    const { grid, start, finish } = this.state;
    const startNode = grid[start.gridId.rowIndex][start.gridId.colIndex];
    const finishNode = grid[finish.gridId.rowIndex][finish.gridId.colIndex];
    const resultOfAStarM = aStarManhatten(grid, startNode, finishNode);
    const y = findShortestPathAStarM(resultOfAStarM[resultOfAStarM.length - 1]);
    this.animateAlgorithm(resultOfAStarM, y);
  };

  runDijkstra = () => {
    const { grid, start, finish } = this.state;
    const startNode = grid[start.gridId.rowIndex][start.gridId.colIndex];
    const finishNode = grid[finish.gridId.rowIndex][finish.gridId.colIndex];
    const resultOfDijkstra = dijkstra(grid, startNode, finishNode);
    const y = findShortestPath(resultOfDijkstra[resultOfDijkstra.length - 1]);
    this.animateAlgorithm(resultOfDijkstra, y);
  };

  // animation
  animateAlgorithm = (visitedNodesInOrder, nodesInShortestPathOrder) => {
    for (let i = 1; i <= visitedNodesInOrder.length - 1; i++) {
      if (i === visitedNodesInOrder.length - 1) {
        setTimeout(() => {
          this.animateShortestPath(nodesInShortestPathOrder);
        }, 5 * i);
        return;
      }
      setTimeout(() => {
        const node = visitedNodesInOrder[i];
        document.getElementById(
          `node-${node.gridId.colIndex}-${node.gridId.rowIndex}`
        ).className = `Node ${
          node.start
            ? 'start'
            : node.finish
            ? 'finish'
            : node.visited
            ? 'visited'
            : ''
        }`;
      }, 5 * i);
    }
  };

  animateShortestPath = (nodesInShortestPathOrder) => {
    for (let i = 1; i < nodesInShortestPathOrder.length - 1; i++) {
      setTimeout(() => {
        const node = nodesInShortestPathOrder[i];
        document.getElementById(
          `node-${node.gridId.colIndex}-${node.gridId.rowIndex}`
        ).className = 'Node path';
      }, 50 * i);
    }
  };

  render() {
    const {
      algorithm,
      fenceToggle,
      finish,
      grid,
      mouseToggle,
      start,
    } = this.state;

    return (
      <Fragment>
        <Header
          algorithm={algorithm}
          ready={start.present && finish.present}
          run={this.run}
          setAlgorithm={this.setAlgorithm}
          fenceToggle={this.fenceToggler}
          reset={this.reset}
        />
        {!start.present && !finish.present ? (
          <Alert variant="primary">Please Choose A Start & End Node</Alert>
        ) : !finish.present ? (
          <Alert variant="primary">Please Choose An End Node</Alert>
        ) : null}

        <Nodes
          grid={grid}
          gridHasStart={start.present}
          gridHasFinish={finish.present}
          fenceToggle={fenceToggle}
          nodeFlag={this.nodeFlag}
          mouseFlag={this.mouseFlag}
          updateNode={this.updateNode}
          mouseToggle={mouseToggle}
        />
      </Fragment>
    );
  }
}
