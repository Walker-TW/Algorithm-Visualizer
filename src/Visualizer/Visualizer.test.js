import React from 'react';

import Visualizer from './Visualizer';
import Node from './Node/Node';

import { shallowToJson } from 'enzyme-to-json';

import { shallow, mount, render } from 'enzyme';

describe('<Visualizer />', () => {
  function toggleFences(wrapper) {
    wrapper
      .find('input[type="checkbox"]')
      .simulate('change', { target: { checked: true } });
  }

  it('renders', () => {
    const wrapper = shallow(<Visualizer />);
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('renders all 1500 Node components', () => {
    const wrapper = mount(<Visualizer />);

    expect(wrapper.find(Node).length).toEqual(1500);
  });

  it('render a node that will change the start state when clicked', () => {
    const wrapper = mount(<Visualizer />);
    const node = wrapper.find(Node).first();

    expect(node.state('start')).toEqual(false);
    node.simulate('mousedown', 'mouseup');
    expect(node.state('start')).toEqual(true);
  });

  it('render a node that will change the finish state when clicked', () => {
    const wrapper = mount(<Visualizer />);
    const node = wrapper.find(Node).first();

    node.simulate('mousedown', 'mouseup');
    expect(wrapper.state().start.present).toEqual(true);
  });

  it('renders 1400 nodes on mobile', () => {
    const wrapperTwo = mount(<Visualizer />);
    expect(wrapperTwo.find(Node).length).toEqual(504);
  });

  it('registers if a mouse is held', () => {
    const wrapper = mount(<Visualizer />);
    const node = wrapper.find(Node).first();
    toggleFences(wrapper);

    node.simulate('mousedown', 'mouseenter');
    expect(wrapper.state('mouseToggle')).toEqual(true);
  });

  it('can add walls', () => {
    const wrapper = mount(<Visualizer />);

    toggleFences(wrapper);

    expect(wrapper.state().fenceToggle).toEqual(true);
  });

  it('can remove walls', () => {
    const wrapper = mount(<Visualizer />);

    toggleFences(wrapper);
    const findnode = wrapper.find('#node-0-0');
    const node = findnode.first();
    [1, 2].forEach(() => {
      node.simulate('click');
    });
    const nodeObj = wrapper.state().grid[0][0];
    expect(nodeObj.fence).toEqual(false);
  });

  // it('sets the algorithm', () => {
  //   const wrapper = mount(<Visualizer />);
  //   const astarLink = wrapper.find('#set-astar');
  //   const dropDown = wrapper.find('#collasible-nav-dropdown');
  //   // console.log(dropDown);
  //   // dropDown.first().simulate('click');
  //   dropDown.first().simulate('mousedown', 'mouseup');
  //   // dropDown.first().simulate('click');
  //   // dropDown[1].simulate('click');
  //   astarLink.simulate('click');
  //   expect(wrapper.state('algorithm')).toEqual('astar');
  // });
});
