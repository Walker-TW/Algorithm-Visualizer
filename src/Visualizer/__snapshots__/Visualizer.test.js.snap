// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Visualizer /> renders 1`] = `
<Fragment>
  <Header
    algorithm=""
    fenceToggle={[Function]}
    ready={false}
    reset={[Function]}
    run={[Function]}
    setAlgorithm={[Function]}
  />
  <Alert
    closeLabel="Close alert"
    show={true}
    transition={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "defaultProps": Object {
          "appear": false,
          "in": false,
          "mountOnEnter": false,
          "timeout": 300,
          "unmountOnExit": false,
        },
        "displayName": "Fade",
        "render": [Function],
      }
    }
    variant="primary"
  >
    Please Choose A Start & End Node
  </Alert>
  <Nodes
    fenceToggle={false}
    grid={Array []}
    gridHasFinish={false}
    gridHasStart={false}
    mouseFlag={[Function]}
    mouseToggle={false}
    nodeFlag={[Function]}
  />
</Fragment>
`;
